### PR TITLE
Extend the signature of imdecodemulti()

### PR DESCRIPTION
**Pull request by kochanczyk**
(Edited after addressing Reviewers' comments.)

Add an argument to `imdecodemulti()` to enable optional selection of pages of multi-page images.

Be default, all pages are decoded. If used, the additional argument may specify a continuous selection of pages to decode.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [X] I agree to contribute to the project under Apache 2 License.
- [X] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
@kochanczyk thanks a lot for the PR. Could you add test like this: https://github.com/opencv/opencv/blob/4.x/modules/imgcodecs/test/test_tiff.cpp#L968.

**Comment by vpisarev** [comment]
@kochanczyk, thanks for the patch! It looks a useful addition. I suggest however not to add extra entry, but add an optional `const Range& range = Range::all()` parameter to the existing function

**Comment by kochanczyk** [comment]
Dear Alex and Vadim, as per your requests, I used cv::Range and provided a separate test. Thank you for your interest and feedback.

File: modules/imgcodecs/test/test_tiff.cpp [hunk 1]
```diff
@@ -1002,6 +1002,59 @@ TEST_P(Imgcodecs_Tiff_Modes, decode_multipage_use_memory_buffer)
...
+    FILE* fp = fopen(filename.c_str(), "rb");
+    ASSERT_TRUE(fp != NULL);
+    fseek(fp, 0, SEEK_END);
+    long pos = ftell(fp);
+
+    std::vector<uchar> buf;
+    buf.resize((size_t)pos);
+    fseek(fp, 0, SEEK_SET);
+    buf.resize(fread(&buf[0], 1, buf.size(), fp));
+    fclose(fp);
```
Comment (asmorkalov) [inline modules/imgcodecs/test/test_tiff.cpp:1029]:
```
size_t file_size = ftell(fp);

std::vector<uchar> buf(file_size);
fseek(fp, 0, SEEK_SET);
size_t actual_read = fread(&buf[0], 1, file_size, fp);
fclose(fp);
ASSERT_EQ(file_size, actual_read);
```

File: modules/imgcodecs/test/test_tiff.cpp [hunk 2]
```diff
@@ -1002,6 +1002,59 @@ TEST_P(Imgcodecs_Tiff_Modes, decode_multipage_use_memory_buffer)
...
+
+            for (int i = range.start; i < range.end; i++)
+            {
+                EXPECT_PRED_FORMAT2(cvtest::MatComparator(0, 0),
+                    pages_from_imread[i],
+                    pages_from_imdecodemulti[i - range.start]);
+            }
+        }
+    }
+}
```
Comment (asmorkalov) [inline modules/imgcodecs/test/test_tiff.cpp:1056]: I do not think that we need all combination here. It's enough to call `imdecodemulti` once with trange like `(1, size-2)` to check not default behaviour. Other cases and tiff specific are indirectly tested with `imreadmulti`.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

File: modules/imgcodecs/test/test_tiff.cpp (hunk 1 above)
Comment (kochanczyk) [reply modules/imgcodecs/test/test_tiff.cpp:1029]: If we are going to revise related code on the occasion (reading binary files would appear in both the tests: `decode_multipage_use_memory_buffer_{all,selected}_pages`), what about replacing:

    size_t file_size = ftell(fp);

    std::vector<uchar> buf(file_size);
    fseek(fp, 0, SEEK_SET);

with:

    size_t file_size = ftell(fp);
    fseek(fp, 0, SEEK_SET);

    std::vector<uchar> buf;
    buf.reserve(file_size);

In the proposed replacement, I/O is not interleaved by memory allocation and we do not unnecessarily initialize a chunk of memory that gets nearly immediately overwritten.

File: modules/imgcodecs/test/test_tiff.cpp (hunk 2 above)
Comment (kochanczyk) [reply modules/imgcodecs/test/test_tiff.cpp:1056]: I'll revise the code as requested, but please note that, strictly speaking, the proposed test does not check all combinations, which would be nonsensical, but all _contiguous subsequences_. If you have a sequence of length _n_, there are _n_ × (_n_ + 1)/2 ~ O(_n_<sup>2</sup>) contiguous subsequences, which for the current _n_ = 6 gives 21.

**Review by asmorkalov — APPROVED** [review]
:+1:

**Review by vpisarev — APPROVED** [review]

File: modules/imgcodecs/src/loadsave.cpp [hunk 3]
```diff
@@ -1095,12 +1095,21 @@ imdecodemulti_(const Mat& buf, int flags, std::vector<Mat>& mats, int start, int
...
     Mat buf = _buf.getMat();
-    return imdecodemulti_(buf, flags, mats, 0, -1);
+    if (range == Range::all())
+    {
+        return imdecodemulti_(buf, flags, mats, 0, -1);
+    }
+    else
+    {
+        CV_Assert( range.start >= 0 );
+        CV_Assert( range.size() > 0 );
```
Comment (opencv-alalek) [inline modules/imgcodecs/src/loadsave.cpp:1110]: CV_CheckGE
CV_CheckGT

**Comment by opencv-alalek** [comment]
Updated skip list with imreadmulti in ABI/API checker on CI.

**Comment by asmorkalov** [comment]
The iOS issue is related to Obj-C/Swift bindings issue:
```
/Users/opencv-cn/GHA-OCV-4/_work/opencv/opencv/ios/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/imgcodecs/Imgcodecs.mm:124:71: error: reference to type 'const cv::Range' could not bind to an lvalue of type 'Range *__strong _Nonnull'
    bool retVal = cv::imdecodemulti(buf.nativeRef, flags, matsVector, range);
                                                                      ^~~~~
In file included from /Users/opencv-cn/GHA-OCV-4/_work/opencv/opencv/ios/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/imgcodecs/Imgcodecs.mm:5:
In file included from /Users/opencv-cn/GHA-OCV-4/_work/opencv/opencv/ios/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/imgcodecs/Imgcodecs.h:8:
/Users/opencv-cn/GHA-OCV-4/_work/opencv/opencv/ios/build/build-armv7-iphoneos/install/include/opencv2/imgcodecs.hpp:342:108: note: passing argument to parameter 'range' here
CV_EXPORTS_W bool imdecodemulti(InputArray buf, int flags, CV_OUT std::vector<Mat>& mats, const cv::Range& range = Range::all());
                                                                                                           ^
1 error generated.
```
@VadimLevin @komakai  could you help with it?

**Comment by VadimLevin** [comment]
The issue lies on generator side, because Obj-C doesn't support default function arguments.

As a workaround you can create 2 functions:

```cpp
CV_EXPORTS_W bool imdecodemulti(InputArray buf, int flags, CV_OUT std::vector<Mat>& mats) {
  return imdecodemulti(buf, flags, mats, Range::all());
}

CV_EXPORTS_W bool imdecodemulti(InputArray buf, int flags, CV_OUT std::vector<Mat>& mats,
                                const cv::Range& range);
```

Or this should be handled on generator side: For each function with default arguments create a trampoline function with and without defaulted argument.

I'm not familiar with Obj-C generator, maybe it is already working that way.

**Comment by komakai** [comment]
@asmorkalov @VadimLevin 
Because the mapping from C++ to Objective-C can never be perfect there are ways to handle situations like this via **gen_dict.json** file
Looking at the config in
https://github.com/opencv/opencv/blob/4.x/modules/core/misc/objc/gen_dict.json
it appears there is an entry already for **cv::Range** but it is incomplete
Possibly adding these lines might fix it
```
            "to_cpp": "%(n)s.nativeRef",
            "from_cpp": "[Range fromNative:%(n)s]"
```

File: modules/core/misc/objc/gen_dict.json [hunk 4]
```diff
@@ -167,7 +167,9 @@
             "from_cpp": "[Point3i fromNative:%(n)s]"
         },
         "Range": {
-            "objc_type": "Range*"
+            "objc_type": "Range*",
+            "to_cpp": "%(n)s.nativeRef",
```
Comment (komakai) [inline modules/core/misc/objc/gen_dict.json:171]: Looks good - let me know if that works OK - if it doesn't then I'll take a closer look

Comment (komakai) [reply modules/core/misc/objc/gen_dict.json:171]: OK - let me take a closer look

**Comment by asmorkalov** [comment]
iOS build:
```
/Users/opencv-cn/GHA-OCV-1/_work/opencv/opencv/ios/build/build-armv7-iphoneos/modules/objc_bindings_generator/ios/gen/objc/imgcodecs/Imgcodecs.mm:124:77: error: property 'nativeRef' not found on object of type 'Range *'
    bool retVal = cv::imdecodemulti(buf.nativeRef, flags, matsVector, range.nativeRef);
                                                                            ^
```

**Comment by komakai** [comment]
@kochanczyk @asmorkalov Can you try with PR #24454 ? That should solve it

**Comment by asmorkalov** [comment]
@komakai Thanks a lot for the quick solution. The related PR has been merged.
@kochanczyk Could you rebase your branch on top of current 4.x to include the patch.

**Comment by asmorkalov** [comment]
Windows:
```
C:/build/precommit_windows64/4.x/opencv/modules/dnn/src/layers/cpu_kernels/softmax_kernels.simd.hpp(276): error C2105: '--' needs l-value (compiling source file C:\build\precommit_windows64\build\modules\dnn\layers\cpu_kernels\softmax_kernels.avx.cpp) [C:\build\precommit_windows64\build\modules\dnn\opencv_dnn_AVX.vcxproj]
```

**Comment by kochanczyk** [comment]
Dear Giles, Alex, and Vadim, thank you very much for all your assistance!  I am happy to see the pull request merged.

**Pull request merged**